src/v2/view-builder/views/DeviceChallengePollView.js

```diff
@@ -13,127 +14,122 @@ const request = (opts) => {
   return $.ajax(ajaxOptions);
 };
 
-const Body = BaseForm.extend({
-  noButtonBar: true,
-
-  className: 'ion-form device-challenge-poll',
-
-  initialize () {
-    BaseForm.prototype.initialize.apply(this, arguments);
-    this.deviceChallengePollRemediation = this.options.appState.getCurrentViewState();
-    this.doChallenge();
-    this.startPolling();
-  },
-
-  remove () {
-    BaseForm.prototype.remove.apply(this, arguments);
-    this.stopPolling();
-  },
-
-  doChallenge () {
-    const deviceChallenge = this.options.appState.get(
-      this.deviceChallengePollRemediation.relatesTo
-    );
-    switch (deviceChallenge.challengeMethod) {
-    case 'LOOPBACK':
-      this.title = loc('signin', 'login');
-      this.add('<div class="spinner"></div>');
-      this.doLoopback(deviceChallenge.domain, deviceChallenge.ports, deviceChallenge.challengeRequest);
-      break;
-    case 'CUSTOM_URI':
-      this.title = 'Verify account access';
-      this.subtitle = 'Launching Okta Verify...';
-      this.add(`
-        If nothing prompts from the browser,  
-        <a href="#" id="launch-ov" class="link">click here</a> to launch Okta Verify, 
-        or make sure Okta Verify is installed.
-      `);
-      this.customURI = deviceChallenge.href;
-      this.doCustomURI();
-      break;
-    }
-  },
-
-  postRender () {
-    BaseForm.prototype.postRender.apply(this, arguments);
-    this.$('#launch-ov').on('click', this.doCustomURI.bind(this));
-  },
-
-  doLoopback (authenticatorDomainUrl = '', ports = [], challengeRequest = '') {
-    let currentPort;
-    let foundPort = false;
-    let countFailedPorts = 0;
-
-    const getAuthenticatorUrl = (path) => {
-      return `${authenticatorDomainUrl}:${currentPort}/${path}`;
-    };
-
-    const checkPort = () => {
-      return request({
-        url: getAuthenticatorUrl('probe'),
-        timeout: 1000 // if authenticator & its probing endpoint exist, it should respond within 1000ms
+const Body = BaseForm.extend(Object.assign(
+  {
+    noButtonBar: true,
+
+    className: 'ion-form device-challenge-poll',
+
+    initialize () {
+      BaseForm.prototype.initialize.apply(this, arguments);
+      this.listenTo(this.model, 'error', this.onPollingFail);
+      this.deviceChallengePollRemediation = this.options.appState.getCurrentViewState();
+      this.doChallenge();
+      this.startDevicePolling();
+    },
+
+    onPollingFail () {
+      this.$('.spinner').hide();
+      this.stopPolling();
+    },
+
+    remove () {
+      BaseForm.prototype.remove.apply(this, arguments);
```

Comment:
when does this get called? When device polling succeeds and you render a new screen?

Reply:
Yes.
This actually needs improvement as I talked to Haisheng before. Currently, SIW is being too aggressively in rerendering the view, it simply compares the before and after IDX response object, as long as there's any diff, model triggers change event thus view rerender. This does not work well at least in polling view. Sometimes the only difference in responses between pollings is only the timestamp, we don't want the view to refresh in that case.